rig-core/src/providers/anthropic/streaming.rs

```diff
@@ -0,0 +1,264 @@
+use futures::{Stream, StreamExt};
+use serde::Deserialize;
+use serde_json::json;
+use std::iter;
+use std::pin::Pin;
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::ReceiverStream;
+
+use super::completion::{CompletionModel, Content, Message, ToolChoice, ToolDefinition, Usage};
+use crate::completion::{CompletionError, CompletionRequest};
+use crate::json_utils::merge_inplace;
+use crate::streaming::{StreamingChoice, StreamingCompletionModel};
+
+#[derive(Debug, Deserialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum StreamingEvent {
+    MessageStart {
+        message: MessageStart,
+    },
+    ContentBlockStart {
+        index: usize,
+        content_block: ContentBlock,
+    },
+    ContentBlockDelta {
+        index: usize,
+        delta: ContentDelta,
+    },
+    ContentBlockStop {
+        index: usize,
+    },
+    MessageDelta {
+        delta: MessageDelta,
+        usage: Usage,
+    },
+    MessageStop,
+    Ping,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct MessageStart {
+    pub id: String,
+    pub role: String,
+    pub content: Vec<Content>,
+    pub model: String,
+    pub stop_reason: Option<String>,
+    pub stop_sequence: Option<String>,
+    pub usage: Usage,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum ContentBlock {
+    Text {
+        text: String,
+    },
+    ToolUse {
+        id: String,
+        name: String,
+        input: serde_json::Value,
+    },
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum ContentDelta {
+    TextDelta { text: String },
+    InputJsonDelta { partial_json: String },
+}
+
+#[derive(Debug, Deserialize)]
+pub struct MessageDelta {
+    pub stop_reason: Option<String>,
+    pub stop_sequence: Option<String>,
+}
+
+#[derive(Default)]
+struct ToolCallState {
+    name: String,
+    id: String,
+    input_json: String,
+}
+
+impl StreamingCompletionModel for CompletionModel {
+    async fn stream(
+        &self,
+        completion_request: CompletionRequest,
+    ) -> Result<
+        Pin<Box<dyn Stream<Item = Result<StreamingChoice, CompletionError>> + Send>>,
+        CompletionError,
+    > {
+        let prompt_with_context = completion_request.prompt_with_context();
+
+        let max_tokens = if let Some(tokens) = completion_request.max_tokens {
+            tokens
+        } else if let Some(tokens) = self.default_max_tokens {
+            tokens
+        } else {
+            return Err(CompletionError::RequestError(
+                "`max_tokens` must be set for Anthropic".into(),
+            ));
```

Comment:
Here instead of returning `Err` you return a `Stream` with an error inside:
```rust
use async_stream::stream;
stream! {
    let error = CompletionError::RequestError("`max_tokens` must be set for Anthropic".into());
    yield(Err(error));
}
```


This way your function will be like this:
```rust
async fn stream(
    &self,
    completion_request: CompletionRequest,
) ->impl TryStream<Ok=StreamingChoice, Err=CompletionError>> + Send {
```